typing/ctype.ml

```diff
@@ -890,30 +890,37 @@ let rec lower_contravariant env var_level visited contra ty =
   in
   if must_visit then begin
     Hashtbl.add visited ty.id contra;
-    let generalize_rec = lower_contravariant env var_level visited in
+    let lower_rec = lower_contravariant env var_level visited in
     match ty.desc with
       Tvar _ -> if contra then set_level ty var_level
-    | Tconstr (path, tyl, abbrev) ->
+    | Tconstr (_, [], _) -> ()
+    | Tconstr (path, tyl, _abbrev) ->
         let variance =
           try (Env.find_type path env).type_variance
           with Not_found ->
             (* See testsuite/tests/typing-missing-cmi-2 for an example *)
             List.map (fun _ -> Variance.may_inv) tyl
         in
-        abbrev := Mnil;
-        List.iter2
-          (fun v t ->
-            if Variance.(mem May_weak v)
-            then generalize_rec true t
-            else generalize_rec contra t)
-          variance tyl
+        if List.for_all ((=) Variance.null) variance then () else
+        begin match !forward_try_expand_once env ty with
+        | ty -> lower_rec contra ty
+        | exception Cannot_expand ->
+          (* abbrev := Mnil;  Should not be needed anymore *)
```

Comment:
I would like to understand this line, and also I think it could be better to not leave this kind of code comments in the source (I feel that nobody except maybe you quite knows why there are here, we are afraid of removing them but they don't make reading the code any easier.)

I looked for `abbrev` usage in the codebase. They are populated by `Btype.memorize_abbrev`, which is itself used by `Btype.subst` only -- so only `subst` creates non-Mnil abbrev caches. `fordward_expand_once` calls `subst`, so it populates the `abbrev` field. The only places which clear those caches are `generalize_structure`, `lower_contraviant` (before this patch) and the global-cleanup functions calling `cleanup_abbrev` (for example `check_abbrev_env` called above).

I have two questions:
1. Why was `abbrev := Mnil` needed before?
2. Why is it "not needed anymore"?

Here is my best shot at answering them:

1. You need to clear the Tconstr abbreviation if you want to change the level (compared to the moment where the expansion was memoized), so structural generalization/lowering functions must clear the abbrev.
2. If you do use `forward_try_expand_once` (as you do in this patch), you want to use the caching provided by the `abbrev` field when you encounter the constructor again, so you do *not* want to clear abbreviations. Besides, `expand_abbrev_gen` is careful to update the levels (I think?), so clearing `abbrev` is unnecessary and also dangerous performance-wise, as it defeats the purpose of expansion caching.

Is that reasoning correct? (Can we remove the commented-out code?)

Reply:
Before a previous fix, `lower_contravariant` was called `generalize_contravariant`, and it was generalizing on its way. As a result this assignment was required. I forgot to remove it when doing that conversion, but it is safe to remove it, and as you remark maybe even required. I tend to be conservative when removing code, but the code is not needed and the comment is correct (this function now only changes the levels of type variables, not of constructor nodes), so it would be possible to remove both.

Reply:
I understand why you don't like this kind of cryptic comments, but they are sometimes helpful when chasing bugs, as they give you some insight of the history of the code. Lots of bugs come from wrong changes.

Reply:
If you want to keep a comment, maybe it could sound a bit more confident. For example:

```
(* Note: we are preserving the `abbrev` cache here to speedup future expansions. This is correct as only the level of type variables is touched in this function, not of type constructors. *)
```